spring-cloud-data-admin/src/main/java/org/springframework/cloud/data/admin/controller/ModuleController.java

```diff
@@ -0,0 +1,158 @@
+import org.springframework.web.bind.annotation.RestController;
+
+/**
+ * Handles all Module related interactions.
+ *
+ * @author Glenn Renfro
+ * @author Mark Fisher
+ * @author Gunnar Hillert
+ * @author Eric Bottard
+ * @author Gary Russell
+ * @author Patrick Peralta
+ */
+@RestController
+@RequestMapping("/modules")
+@ExposesResourceFor(ModuleRegistrationResource.class)
+public class ModuleController {
+
+	private final Assembler moduleAssembler = new Assembler();
+
+	private final ModuleRegistry registry;
+
+	@Autowired
+	public ModuleController(ModuleRegistry registry) {
+		this.registry = registry;
+	}
+
+	/**
+	 * List module registrations.
+	 */
+	@RequestMapping(method = RequestMethod.GET)
+	@ResponseStatus(HttpStatus.OK)
+	public PagedResources<? extends ModuleRegistrationResource> list(Pageable pageable,
+			PagedResourcesAssembler<ModuleRegistration> assembler,
+			@RequestParam(value = "type", required = false) ModuleType type,
+			@RequestParam(value = "detailed", defaultValue = "false") boolean detailed) {
+
+		List<ModuleRegistration> list = new ArrayList<>(registry.findAll());
+		if (type != null) {
+			for (Iterator<ModuleRegistration> iterator = list.iterator(); iterator.hasNext();) {
+				if (iterator.next().getType() != type) {
+					iterator.remove();
+				}
+			}
+		}
+
+		// modules have a "natural" sort order of type followed by name,
+		// use that instead of what "pageable" indicates as a sort order
+		Collections.sort(list);
+		return assembler.toResource(new PageImpl<>(list), moduleAssembler);
+	}
+
+	/**
+	 * Retrieve detailed information about a particular module.
+	 *
+	 * @param type  module type
+	 * @param name  module name
+	 * @return detailed module information
+	 */
+	@RequestMapping(value = "/{type}/{name}", method = RequestMethod.GET)
+	@ResponseStatus(HttpStatus.OK)
+	public DetailedModuleRegistrationResource info(
+			@PathVariable("type") ModuleType type,
+			@PathVariable("name") String name) {
+		ModuleRegistration registration = registry.find(name, type);
+
+		return (registration == null ? null :
+				new DetailedModuleRegistrationResource(moduleAssembler.toResource(registration)));
+	}
+
+	/**
+	 * Register a module name and type with its Maven coordinates.
+	 *
+	 * @param type  module type
+	 * @param name  module name
+	 * @param coordinates  Maven coordinates for the module artifact
+	 * @param force if {@code true}, overwrites a pre-existing registration
+	 */
+	@RequestMapping(value = "/{type}/{name}", method = RequestMethod.POST)
+	@ResponseStatus(HttpStatus.CREATED)
+	public void register(
+			@PathVariable("type") ModuleType type,
+			@PathVariable("name") String name,
+			@RequestParam("coordinates") String coordinates,
+			@RequestParam(value = "force", defaultValue = "false") boolean force) {
+		if (!force && registry.find(name, type) != null) {
+			return;
+		}
+		registry.save(new ModuleRegistration(name, type, ModuleCoordinates.parse(coordinates)));
```

Comment:
If a stream is redeployed using a module that has been explicitly updated and that update causes a failure, then the redeploy itself will fail. That should be expected. I don't think we should couple the registry to the runtime state of streams. By requiring "force" to override any existing registry entry, we are already requiring the user to think twice about updates to modules that *might* be in use already.